nixops/state/json_file.py

```diff
@@ -0,0 +1,356 @@
+                json.dump(initial_db, f)
+                f.close()
+
+    def schema_version(self):
+        version = self.read()["schemaVersion"]
+        if version is None:
+            raise "illegal datafile" # TODO: proper exception
+        else:
+            return version
+
+class JsonFile(object):
+    """NixOps state file."""
+
+    def __init__(self, json_file):
+        self.file_path = json_file
+
+        if os.path.splitext(json_file)[1] not in ['.json']:
+            raise Exception("state file ‘{0}’ should have extension ‘.json’".format(json_file))
+
+        self.db = TransactionalJsonFile(json_file)
+
+        # Check that we're not using a new DB schema version.
+        with self.db:
+            version = self.db.schema_version()
+            if version > 0:
+               raise Exception("this NixOps version is too old to deal with JSON schema version {0}".format(version))
+
+    def close(self):
+        pass
+
+    ###############################################################################################
+    ## Deployment
+
+    def query_deployments(self):
+        """Return the UUIDs of all deployments in the database."""
+
+        return self.db.read()["deployments"].keys()
+
+    def get_all_deployments(self):
+        """Return Deployment objects for every deployment in the database."""
+        uuids = self.query_deployments()
+        res = []
+        for uuid in uuids:
+            try:
+                res.append(self.open_deployment(uuid=uuid))
+            except nixops.deployment.UnknownBackend as e:
+                sys.stderr.write("skipping deployment ‘{0}’: {1}\n".format(uuid, str(e)))
+        return res
+
+    def _find_deployment(self, uuid=None):
+        all_deployments = self.db.read()["deployments"]
+
+        found = []
+
+        # if nothing exists no reason to check for things
+        if not all_deployments:
+            return None
+
+        if not uuid:
+            found = filter(lambda(id): all_deployments[id]["attributes"].get("name"), all_deployments)
+        else:
+            found = filter(lambda(id): id == uuid, all_deployments)
+            if not found:
+                found = filter(lambda(id): all_deployments[id]["attributes"].get("name") == uuid, all_deployments)
+            if not found:
+                found = filter(lambda(id): id.startswith(uuid), all_deployments)
+
+        if not found:
+            return None
+
+        if len(found) > 1:
+            if uuid:
+                raise Exception("state file contains multiple deployments with the same name, so you should specify one using its UUID")
+            else:
+                raise Exception("state file contains multiple deployments, so you should specify which one to use using ‘-d’, or set the environment variable NIXOPS_DEPLOYMENT")
+        return nixops.deployment.Deployment(self, found[0], sys.stderr)
+
+    def open_deployment(self, uuid=None):
+        """Open an existing deployment."""
```

Comment:
What's the return "type" of calling this? E.g. it could return some abstract file pointer and abstract deployment resource anything, if you just look at this method's doc string.

Reply:
it's `None` or `nixops.deployment.Deployment` object. 

Is the `Deployment` class abstract (i.e. it's inherited by someone other class)? Looking at the source code - no, it's not.